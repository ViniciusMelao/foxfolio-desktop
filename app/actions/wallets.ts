import { Wallet } from '../modules/wallets.types';
import { Action, Dispatch, ThunkAction } from './actions.types';
import { requestTickerUpdate } from './ticker';

export function addWallet(wallet: Wallet): ThunkAction {
  return (dispatch: Dispatch) => {
    dispatch(requestTickerUpdate([wallet.currency]));
    dispatch({
      type: 'ADD_WALLET',
      wallet,
    });
  };
}

export function editWallet(wallet: Wallet, newWallet: Wallet): ThunkAction {
  return (dispatch: Dispatch) => {
    dispatch(requestTickerUpdate([wallet.currency]));
    dispatch({
      type: 'EDIT_WALLET',
      wallet,
      newWallet,
    });
  };
}

export function deleteWallet(wallet: Wallet): Action {
  return {
    type: 'DELETE_WALLET',
    wallet,
  };
}
