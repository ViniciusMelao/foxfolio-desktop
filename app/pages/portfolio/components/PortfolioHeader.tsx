import { Grid, Typography } from '@material-ui/core';
import React from 'react';
import { SettingsType } from '../../../modules/settings.types';
import { PortfolioChange, PortfolioSum } from '../types/portfolio.types';
import { PriceChangeText } from './PriceChangeText';

interface Props {
  settings: SettingsType;
  sum: PortfolioSum;
  change: PortfolioChange;
}

export const PortfolioHeader = ({ settings, sum, change }: Props) => {
  return (
    <Grid container>
      <Grid item xs={2} />
      <Grid item xs={4}>
        <Typography variant="title">
          {sum.fiat.toPrecision(5)} {settings.fiatCurrency} |&nbsp;
          <PriceChangeText change={change.fiat} />
        </Typography>
      </Grid>
      <Grid item xs={4}>
        <Typography variant="title" style={{ paddingLeft: 20 }}>
          {sum.crypto.toPrecision(5)} {settings.cryptoCurrency} |{' '}
          <PriceChangeText change={change.crypto} />
        </Typography>
      </Grid>
    </Grid>
  );
};
