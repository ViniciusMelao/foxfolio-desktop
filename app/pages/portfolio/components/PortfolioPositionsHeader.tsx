import { Button, Card, CardContent, Grid, Typography } from '@material-ui/core';
import { StyleRulesCallback, Theme, withStyles } from '@material-ui/core/styles';
import React from 'react';

const styles: StyleRulesCallback = (theme: Theme) => ({
  root: {
    display: 'flex',
    alignItems: 'center',
    paddingTop: '12px !important',
    paddingBottom: '12px !important',
  },
  avatar: {
    flex: '0 0 auto',
    width: 40,
    marginRight: theme.spacing.unit * 2,
  },
  content: {
    flex: '1 1 auto',
    marginRight: theme.spacing.unit * 2,
  },
  collapseButton: {
    flex: '0 0 auto',
    width: 48,
  },
  right: {
    textAlign: 'right',
  },
  interval: {
    fontSize: 12,
    color: 'grey',
  },
});

export const PortfolioPositionHeader = withStyles(styles)(({ classes }) => (
  <Card square>
    <CardContent className={classes.root}>
      <div className={classes.avatar} />
      <div className={classes.content}>
        <Grid container>
          <Grid item xs={4}>
            <Typography variant="subheading">Coin</Typography>
          </Grid>
          <Grid item xs={3} className={classes.right}>
            <Typography variant="subheading">Quantity</Typography>
          </Grid>
          <Grid item xs={3} className={classes.right}>
            <Typography variant="subheading">Price</Typography>
          </Grid>
          <Grid item xs={2} className={classes.right}>
            <Typography variant="subheading">Change</Typography>
          </Grid>
        </Grid>
      </div>
      <div className={classes.collapseButton} />
    </CardContent>
  </Card>
));
