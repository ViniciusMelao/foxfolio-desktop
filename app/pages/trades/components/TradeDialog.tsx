import {
  Button,
  Dialog,
  DialogActions,
  DialogContent,
  DialogTitle,
  Grid,
  MenuItem,
  TextField,
  WithStyles,
} from '@material-ui/core';
import { StyleRules, withStyles } from '@material-ui/core/styles';
import dayjs from 'dayjs';
import React, { Component, FormEvent } from 'react';
import { Trade } from '../../../reducers/exchanges.types';
import { DialogConfig } from '../../exchange/components/ExchangeDialog';

interface Props {
  open: boolean;
  onClose: () => void;
  saveTrade: (trade: Trade) => void;
  config?: DialogConfig<Trade>;
}

type State = Trade;

const emptyTrade: Trade = {
  id: '',
  amount: 0,
  cost: 0,
  datetime: new Date(),
  price: 0,
  side: 'buy',
  symbol: '/',
  timestamp: Date.now(),
};

const styles: StyleRules = {
  paper: {
    width: 500,
  },
};

export const TradeDialog = withStyles(styles)(
  class extends Component<Props & WithStyles, State> {
    public state: State = emptyTrade;

    public componentWillReceiveProps(nextProps: Props): void {
      if (this.props.open !== nextProps.open && nextProps.config) {
        if (nextProps.config.action === 'edit') {
          return this.setState(nextProps.config.item);
        }
        this.setState(emptyTrade);
      }
    }

    public handleChange = (name: keyof State) => (event: React.ChangeEvent<HTMLInputElement>) => {
      this.setState({
        ...this.state,
        [name]: event.target.value,
      });
    };

    public handleSymbolChange = (part: 'from' | 'to') => (
      event: React.ChangeEvent<HTMLInputElement>
    ) => {
      const value = event.target.value;
      this.setState({
        ...this.state,
        symbol:
          part === 'from'
            ? `${value}/${this.state.symbol.split('/' || '')[1]}`
            : `${this.state.symbol.split('/')[0] || ''}/${value}`,
      });
    };

    public handleNumericChange = (name: keyof State) => (
      event: React.ChangeEvent<HTMLInputElement>
    ) => {
      this.setState({
        ...this.state,
        [name]: parseFloat(event.target.value),
      });
    };

    public handleDateChange = () => (event: React.ChangeEvent<HTMLInputElement>) => {
      this.setState({
        ...this.state,
        datetime: dayjs(event.target.value).toDate(),
        timestamp: dayjs(event.target.value).valueOf(),
      });
    };

    public handleSubmit = (event: FormEvent<HTMLFormElement>) => {
      event.preventDefault();
      this.props.saveTrade({ ...this.state, cost: this.state.price * this.state.amount });
    };

    public createDialog(config: DialogConfig<Trade>) {
      return (
        <div>
          <DialogTitle>{config.action === 'add' ? 'Add' : 'Edit '} exchange </DialogTitle>
          <DialogContent>
            <form autoComplete="off" onSubmit={this.handleSubmit}>
              <div style={{ minHeight: 250 }}>
                <TextField
                  label="Side"
                  id="side"
                  value={this.state.side}
                  onChange={this.handleChange('side')}
                  fullWidth
                  margin="normal"
                  select
                >
                  <MenuItem value="buy">buy</MenuItem>
                  <MenuItem value="sell">sell</MenuItem>
                </TextField>
                <Grid container>
                  <Grid item xs={6}>
                    <TextField
                      label="From"
                      id="from"
                      value={this.state.symbol.split('/')[0] || ''}
                      onChange={this.handleSymbolChange('from')}
                      fullWidth
                      margin="normal"
                    />
                  </Grid>
                  <Grid item xs={6}>
                    <TextField
                      label="To"
                      id="to"
                      value={this.state.symbol.split('/')[1] || ''}
                      onChange={this.handleSymbolChange('to')}
                      fullWidth
                      margin="normal"
                    />
                  </Grid>
                </Grid>
                <TextField
                  label="Datetime"
                  id="datetime"
                  value={dayjs(this.state.datetime).toISOString()}
                  onChange={this.handleDateChange()}
                  fullWidth
                  margin="normal"
                />
                <TextField
                  label="Amount"
                  id="amount"
                  type="number"
                  value={this.state.amount}
                  onChange={this.handleNumericChange('amount')}
                  fullWidth
                  margin="normal"
                />
                <TextField
                  label="Price"
                  id="price"
                  type="number"
                  value={this.state.price}
                  onChange={this.handleNumericChange('price')}
                  fullWidth
                  margin="normal"
                />
              </div>
              <DialogActions>
                <Button onClick={this.props.onClose} color="primary">
                  Cancel
                </Button>
                <Button type="submit" color="primary">
                  Ok
                </Button>
              </DialogActions>
            </form>
          </DialogContent>
        </div>
      );
    }

    public render() {
      const { open, onClose, config, classes } = this.props;

      return (
        <Dialog open={open} onClose={onClose} classes={{ paper: classes.paper }}>
          {config ? this.createDialog(config) : ''}
        </Dialog>
      );
    }
  }
);
