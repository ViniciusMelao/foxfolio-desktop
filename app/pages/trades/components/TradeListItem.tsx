import { WithStyles } from '@material-ui/core';
import green from '@material-ui/core/colors/green';
import red from '@material-ui/core/colors/red';
import Grid from '@material-ui/core/Grid';
import { StyleRulesCallback } from '@material-ui/core/styles';
import withStyles from '@material-ui/core/styles/withStyles';
import Typography from '@material-ui/core/Typography';
import dayjs from 'dayjs';
import * as React from 'react';
import { CurrencyAvatar } from '../../../components/CurrencyAvatar';
import { getPriceForTime, getTickerEntry } from '../../../helpers/ticker';
import { Coinlist } from '../../../modules/coinlist.types';
import { Trade } from '../../../modules/exchanges.types';
import { SettingsType } from '../../../modules/settings.types';
import { PricesForTime, Ticker } from '../../../modules/ticker.types';

const styles: StyleRulesCallback = theme => ({
  avatar: {
    flex: '0 0 auto',
    marginRight: theme.spacing.unit * 2,
  },
  content: {
    flex: '1 1 auto',
  },
  right: {
    textAlign: 'right',
  },
  positive: {
    color: green[500],
  },
  negative: {
    color: red[500],
  },
});

export interface Props {
  coinlist: Coinlist;
  trade: Trade;
  settings: SettingsType;
  ticker: Ticker;
  pricesForTime: PricesForTime;
}

export const TradeListItem = withStyles(styles)(
  ({ classes, coinlist, ticker, trade, settings, pricesForTime }: Props & WithStyles) => {
    const asset = trade.symbol.split('/')[0];
    const currency = trade.symbol.split('/')[1];
    const currentPrice = getTickerEntry(ticker, asset, settings.fiatCurrency).PRICE;
    const historicalPrice = getPriceForTime(
      pricesForTime,
      currency,
      settings.fiatCurrency,
      trade.timestamp
    );

    const currentValue = currentPrice * trade.amount;
    const historicalValue = trade.amount * trade.price * historicalPrice;
    const change = currentValue / historicalValue * 100 - 100;
    return (
      <React.Fragment>
        <div className={classes.avatar}>
          <CurrencyAvatar asset={asset} coinlist={coinlist} />
        </div>
        <div className={classes.content}>
          <Grid container>
            <Grid item xs={2}>
              <Typography variant="body1">{dayjs(trade.timestamp).format('MMM DD, YYYY')}</Typography>
            </Grid>
            <Grid item xs={4} className={classes.right}>
              <Typography variant="body1">
                {trade.amount} x {trade.price.toFixed(6)} {settings.fiatCurrency} ={' '}
                {historicalValue.toFixed(2)} {settings.fiatCurrency}
              </Typography>
            </Grid>
            <Grid item xs={2} className={classes.right}>
              <Typography
                variant="body1"
                className={change >= 0 ? classes.positive : classes.negative}
              >
                {change >= 0 ? '+' : ''}
                {change.toFixed(2)}%
              </Typography>
            </Grid>
            <Grid item xs={2} className={classes.right}>
              <Typography variant="body1">
                {(currentPrice * trade.amount).toFixed(2)} {settings.fiatCurrency}
              </Typography>
            </Grid>
            <Grid item xs={2} className={classes.right}>
              <Typography variant="body1">
                {(currentValue - historicalValue).toFixed(2)} {settings.fiatCurrency}
              </Typography>
            </Grid>
          </Grid>
        </div>
      </React.Fragment>
    );
  }
);
