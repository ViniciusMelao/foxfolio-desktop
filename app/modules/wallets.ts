import _ from 'lodash';
import { Action } from '../actions/actions.types';
import { Wallet } from './wallets.types';

export function wallets(state: Wallet[] = [], action: Action): Wallet[] {
  switch (action.type) {
    case 'ADD_WALLET':
      return [...state, action.wallet];
    case 'EDIT_WALLET':
      return [..._.reject(state, item => _.isEqual(item, action.wallet)), action.newWallet];
    case 'DELETE_WALLET':
      return _.reject(state, item => _.isEqual(item, action.wallet));
    default:
      return state;
  }
}
