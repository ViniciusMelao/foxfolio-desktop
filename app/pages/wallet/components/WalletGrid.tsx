import { Button, createStyles, Grid, Typography, WithStyles } from '@material-ui/core';
import { StyleRulesCallback, withStyles } from '@material-ui/core/styles';
import { Add } from '@material-ui/icons';
import React, { Component } from 'react';
import { Coinlist } from '../../../modules/coinlist.types';
import { Wallet } from '../../../modules/wallets.types';
import { WalletCard } from './WalletCard';
import WalletDialog from './WalletDialog';

const styles: StyleRulesCallback = theme =>
  createStyles({
    button: {
      margin: theme.spacing.unit,
      color: theme.palette.background.default,
      position: 'fixed',
      right: 40,
      bottom: 40,
    },
  });

interface Props extends WithStyles<typeof styles> {
  coinlist: Coinlist;
  wallets: Wallet[];
  addWallet: (source: Wallet) => any;
  editWallet: (oldWallet: Wallet, newWallet: Wallet) => any;
  deleteWallet: (source: Wallet) => any;
}

interface State {
  open: boolean;
  isNew: boolean;
  currentWallet: Wallet;
}

export const WalletGrid = withStyles(styles)(
  class extends Component<Props, State> {
    public state = {
      open: false,
      isNew: true,
      currentWallet: { currency: '', address: '', quantity: 0 },
    };

    public addDialog = () => {
      this.setState({
        open: true,
        isNew: true,
        currentWallet: { currency: '', address: '', quantity: 0 },
      });
    };

    public editDialog = (wallet: Wallet) => {
      this.setState({ open: true, isNew: false, currentWallet: wallet });
    };

    public closeDialog = () => {
      this.setState({ open: false });
    };

    public saveDialog = (wallet: Wallet) => {
      if (this.state.isNew) {
        this.props.addWallet(wallet);
      } else {
        this.props.editWallet(this.state.currentWallet, wallet);
      }
      this.closeDialog();
    };

    public render() {
      const { wallets, classes, coinlist, deleteWallet } = this.props;

      return (
        <div className="container">
          <Typography variant="headline">Wallets</Typography>
          <Grid container spacing={16}>
            {wallets.sort((a, b) => a.address.localeCompare(b.address)).map(wallet => (
              <Grid item key={wallet.currency + wallet.address} sm={12} md={8} lg={6}>
                <WalletCard
                  wallet={wallet}
                  onEdit={() => this.editDialog(wallet)}
                  onDelete={() => deleteWallet(wallet)}
                  coinlist={coinlist}
                />
              </Grid>
            ))}
          </Grid>
          <Button
            variant="fab"
            color="primary"
            aria-label="add"
            className={classes.button}
            onClick={this.addDialog}
          >
            <Add />
          </Button>
          <WalletDialog
            open={this.state.open}
            wallet={this.state.currentWallet}
            close={this.closeDialog}
            save={this.saveDialog}
            coinlist={coinlist}
          />
        </div>
      );
    }
  }
);
