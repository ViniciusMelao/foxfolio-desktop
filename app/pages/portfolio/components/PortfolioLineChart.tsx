import { LinearProgress, withTheme } from '@material-ui/core';
import { WithTheme } from '@material-ui/core/styles/withTheme';
import dayjs from 'dayjs';
import React, { Component } from 'react';
import { connect, MapStateToProps } from 'react-redux';
import {
  Area,
  AreaChart,
  CartesianGrid,
  ResponsiveContainer,
  Tooltip,
  XAxis,
  YAxis,
} from 'recharts';
import { bindActionCreators } from 'redux';

import * as _ from 'lodash';
import { Dispatch } from '../../../actions/actions.types';
import { getHistoryEntry } from '../../../helpers/ticker';
import { GlobalState } from '../../../modules';
import * as HistoryActions from '../../../modules/history';
import { History } from '../../../modules/history.types';
import { SettingsType } from '../../../modules/settings.types';
import { getHistory, getSettings } from '../../../selectors/selectGlobalState';
import { Balances } from '../types/portfolio.types';

interface StateProps {
  history: History;
  settings: SettingsType;
}

interface DispatchProps {
  requestHistoryForAll: (fsyms: string[], tsym: string) => void;
}

interface OwnProps {
  portfolio: Balances;
  sum: number;
}

export const ThemedPortfolioLineChart = withTheme()(
  class extends Component<StateProps & DispatchProps & OwnProps & WithTheme> {
    public componentWillMount() {
      const { requestHistoryForAll, portfolio, settings } = this.props;
      requestHistoryForAll(Object.keys(portfolio), settings.cryptoCurrency);
    }

    public render() {
      const { history, portfolio, sum, settings, theme } = this.props;

      const data = calculateHistoryForPortfolio(history, portfolio, settings.cryptoCurrency, sum);
      if (data.length > 0) {
        return (
          <ResponsiveContainer height={300}>
            <AreaChart data={data} margin={{ left: 5, top: 5, bottom: 0, right: 15 }}>
              <Area
                stroke={theme.palette.text.secondary}
                fill={theme.palette.divider}
                dataKey="value"
                isAnimationActive={false}
              />
              <XAxis
                dataKey="time"
                tick={{ fill: theme.palette.text.secondary, strokeWidth: 0 }}
                tickFormatter={tick => dayjs.unix(tick).format('HH:mm')}
              />
              <YAxis
                interval="preserveStartEnd"
                tick={{ fill: theme.palette.text.secondary, strokeWidth: 0 }}
                tickFormatter={tick => tick.toPrecision(3)}
                domain={[dataMin => dataMin - 0.02 * dataMin, dataMax => dataMax + 0.02 * dataMax]}
              />
              <CartesianGrid stroke={theme.palette.divider} strokeDasharray="3 3" />
              <Tooltip
                itemStyle={{ color: theme.palette.text.secondary }}
                wrapperStyle={{ backgroundColor: theme.palette.background.paper }}
                labelFormatter={() => ''}
              />
            </AreaChart>
          </ResponsiveContainer>
        );
      } else {
        return <LinearProgress />;
      }
    }
  }
);

const calculateHistoryForPortfolio = (
  history: History,
  portfolio: Balances,
  tsym: string,
  sum: number
): Array<{ time: number; value: number }> => {
  _.mapValues(portfolio, balance => balance / sum);
  const mergedHistory: { [time: string]: number } = {};
  Object.keys(portfolio).forEach(fsym => {
    const data = getHistoryEntry(history, fsym, tsym).data;
    data.forEach(
      value =>
        (mergedHistory[value.time] =
          (mergedHistory[value.time] || 0) + value.close * portfolio[fsym])
    );
  });
  return _.entries(mergedHistory).map(pair => ({
    time: parseInt(pair[0], 10),
    value: Math.floor((pair[1] + (portfolio[tsym] || 0)) * 1e6) / 1e6,
  }));
};

const mapStateToProps: MapStateToProps<StateProps, OwnProps, GlobalState> = state => ({
  history: getHistory(state),
  settings: getSettings(state),
});
const mapActionsToProps = (dispatch: Dispatch) => bindActionCreators(HistoryActions, dispatch);

export const PortfolioLineChart = connect<StateProps, DispatchProps, OwnProps, GlobalState>(
  mapStateToProps,
  mapActionsToProps
)(ThemedPortfolioLineChart);
