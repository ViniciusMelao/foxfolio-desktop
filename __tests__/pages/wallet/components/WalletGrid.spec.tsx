import { createShallow } from '@material-ui/core/test-utils';
import * as React from 'react';
import { Wallet } from '../../../../app/modules/wallets.types';
import { WalletGrid } from '../../../../app/pages/wallet/components/WalletGrid';

let shallow;

beforeAll(() => {
  shallow = createShallow();
});

let wallets: Wallet[];

beforeEach(() => {
  wallets = [{ address: '', currency: 'BTC', quantity: 1 }];
});

test('Render wallet grid', () => {
  const coinlist = {};
  const addWallet = jest.fn();
  const editWallet = jest.fn();
  const deleteWallet = jest.fn();

  const wrapper = shallow(
    <WalletGrid
      coinlist={coinlist}
      wallets={wallets}
      addWallet={addWallet}
      editWallet={editWallet}
      deleteWallet={deleteWallet}
    />
  ).dive();
  expect(wrapper).toMatchSnapshot();
});
