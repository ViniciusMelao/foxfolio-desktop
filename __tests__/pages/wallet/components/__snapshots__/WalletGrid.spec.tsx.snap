// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Render wallet grid 1`] = `
<div
  className="container"
>
  <WithStyles(Typography)
    variant="headline"
  >
    Wallets
  </WithStyles(Typography)>
  <WithStyles(Grid)
    container={true}
    spacing={16}
  >
    <WithStyles(Grid)
      item={true}
      key="btc"
      lg={6}
      md={8}
      sm={12}
    >
      <WithStyles(class_1)
        coinlist={Object {}}
        onDelete={[Function]}
        onEdit={[Function]}
        wallet={
          Object {
            "currency": "BTC",
            "id": "btc",
            "quantity": 1,
          }
        }
      />
    </WithStyles(Grid)>
  </WithStyles(Grid)>
  <WithStyles(Button)
    aria-label="add"
    className="class_1-button-1"
    color="primary"
    onClick={[Function]}
    variant="fab"
  >
    <pure(AddIcon) />
  </WithStyles(Button)>
  <WalletDialog
    close={[Function]}
    coinlist={Object {}}
    open={false}
    save={[Function]}
    wallet={
      Object {
        "currency": "",
        "id": "",
        "quantity": 0,
      }
    }
  />
</div>
`;
